runtime/common/metering.go

```diff
@@ -217,121 +232,170 @@ func max(a, b int) int {
 	return b
 }
 
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
 const bigIntWordSize = int(unsafe.Sizeof(big.Word(0)))
 
+var bigIntWordSizeAsBig = big.NewInt(int64(bigIntWordSize))
+
 func BigIntByteLength(v *big.Int) int {
-	// NOTE: big.Int.Bits() actually returns bytes:
-	// []big.Word, where big.Word = uint
+	// NOTE: big.Int.Bits() actually returns a slice of words,
+	// []big.Word, where big.Word = uint,
+	// NOT a slice of bytes!
 	return len(v.Bits()) * bigIntWordSize
 }
 
 func NewPlusBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
+	maxWordLength := max(
+		len(a.Bits()),
+		len(b.Bits()),
+	)
 	return NewBigIntMemoryUsage(
-		// TODO: https://github.com/dapperlabs/cadence-private-issues/issues/32
-		max(
-			BigIntByteLength(a),
-			BigIntByteLength(b),
-		) + bigIntWordSize,
+		(maxWordLength + 4) *
+			bigIntWordSize,
 	)
 }
 
 func NewMinusBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
+	maxWordLength := max(
+		len(a.Bits()),
+		len(b.Bits()),
+	)
 	return NewBigIntMemoryUsage(
-		// TODO: https://github.com/dapperlabs/cadence-private-issues/issues/32
-		max(
-			BigIntByteLength(a),
-			BigIntByteLength(b),
-		),
+		(maxWordLength + 5) *
+			bigIntWordSize,
 	)
 }
 
 func NewMulBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
-	return NewBigIntMemoryUsage(
-		// TODO: https://github.com/dapperlabs/cadence-private-issues/issues/32
-		BigIntByteLength(a) +
-			BigIntByteLength(b),
+	aWordLength := len(a.Bits())
+	bWordLength := len(b.Bits())
+	minWordLength := min(
+		aWordLength,
+		bWordLength,
 	)
-}
+	wordLengthSum := aWordLength + bWordLength
 
-func NewTypeMemoryUsage(staticTypeAsString string) MemoryUsage {
-	return MemoryUsage{
-		Kind:   MemoryKindTypeValue,
-		Amount: uint64(len(staticTypeAsString)),
+	var resultWordLength int
+	if minWordLength <= 40 {
+		resultWordLength = wordLengthSum + 4
+	} else {
+		resultWordLength = 3*minWordLength + max(6*minWordLength, wordLengthSum) + 8
 	}
+	return NewBigIntMemoryUsage(
+		resultWordLength * bigIntWordSize,
+	)
 }
 
-func NewCharacterMemoryUsage(length int) MemoryUsage {
-	return MemoryUsage{
-		Kind:   MemoryKindCharacter,
-		Amount: uint64(length),
-	}
-}
+var bigOne = big.NewInt(1)
+var bigOneHundred = big.NewInt(100)
 
 func NewModBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
+	aWordLength := len(a.Bits())
+	bWordLength := len(b.Bits())
+
+	var resultWordLength int
+	if a.Cmp(b) < 0 || b.Cmp(bigOne) == 0 {
+		resultWordLength = aWordLength
+	} else if b.Cmp(bigOneHundred) < 0 {
+		resultWordLength = aWordLength - bWordLength + 5
+	} else {
+		recursionCost := int(unsafe.Sizeof(uintptr(0))) +
+			9*bWordLength +
+			int(math.Floor(float64(aWordLength)/float64(bWordLength))) + 12
+		recursionDepth := 2 * b.BitLen()
+		resultWordLength = 3*bWordLength + 4 + recursionCost*recursionDepth
+	}
```

Comment:
Might be useful to have a comment with the used equations for reference purposes

Reply:
Good idea, added in [c094707](https://github.com/onflow/cadence/pull/1590/commits/c09470796f1fe6e1eec16cf8ab37c9724ac76e52)